tests/test_coroutine.rs

```diff
@@ -0,0 +1,9 @@
+#![cfg(feature = "async")]
+
+use pyo3::prelude::*;
+
+#[path = "../src/tests/common.rs"]
+mod common;
+
+#[pyfunction]
+async fn _async_fn_is_supported() {}
```

Comment:
So I suppose that the challenge now is that even if we've gotten this far, we can't await on other Rust stuff that requires an async runtime because there isn't a Rust async runtime running?

I wonder, should we consider adding `impl Future for PyAny`, with the item being `PyResult<PyAny>`? That way, I guess we can start awaiting other Python objects, assuming a Python runtime is running...

(I haven't thought that second part through to completion.)

Reply:
> we can't await on other Rust stuff that requires an async runtime because there isn't a Rust async runtime wrunning

Actually, you don't need a running Rust runtime to have something to await. `poll_fn`, `AtomicWaker` `thread::spawn`, etc. can do the job. I haven't write more than an empty test simply because I haven't taken the time yet.

> I wonder, should we consider adding impl Future for PyAny, with the item being PyResult<PyAny>?

Things are quite more complex than this when it comes to await Python awaitable. Contrary to Rust, Python coroutines should always be run to completion (no RAII-style future that can be dropped in the middle of the execution), that's why they should always be executed in a spawned task. So you should only await `asyncio.Future` or `asyncio.Task`, see pyo3-async [`FutureWrapper`](https://docs.rs/pyo3-async/latest/pyo3_async/asyncio/struct.FutureWrapper.html) for example.
(pyo3-async also provides [`AwaitableWrapper`](https://docs.rs/pyo3-async/latest/pyo3_async/asyncio/struct.AwaitableWrapper.html) to run manually an awaitable, but as mentioned above, this is not a good thing to do and I planned to remove it).

But that get event more complicated if you want to directly spawn the asyncio task in Rust. pyo3-asyncio does that, but it requires to keep the event loop reference all along, and to acquire the GIL at every spawn, and it doest support task group and so on. I've thought about a way to improve this a bit, with a "spawner" (which could use a task group) task running in asyncio background and fed by an mpsc; writing to the mpsc would require to acquire the GIL only when the receiver is awaiting, so not always, and being in the same event loop thread make possible to avoid `call_soon_threadsafe` call for callback registration. You can even go further with some `await asyncio.sleep(0)`-based spin loop on mpsc receiver side in order to avoid awaiting, and thus to avoid GIL acquiring on mpsc writer thread for the wakeup.
But that seems to be a little too much complex to be integrated in PyO3, even more considering exception handling in tasks if you wanna play with task groups.

I think that awaiting Python objects should be a completely seperate topic for the moment.

Reply:
Thank you, that's a lot of very good reasons why we don't want to do that yet 👍